src/plugins/rpd/rpdplugin.cpp

```diff
@@ -0,0 +1,319 @@
+        if (value.canConvert<double>()) {
+            mapJson.insert(key, value.toDouble());
+            continue;
+        }
+
+        if (value.canConvert<QString>()) {
+            mapJson.insert(key, value.toString());
+            continue;
+        }
+
+        mError = tr("Dont know what to do with property (%1)").arg(key);
+        return false;
+    }
+
+    for (Tiled::Layer *layer : map->layers()) {
+        if (layer->layerType() == Tiled::Layer::TileLayerType
+            && !insertTilesetFile(*layer, QString("tiles_") + layer->name(), mapJson)) {
+            return false;
+        }
+
+        if (layer->name() == "logic") {
+            QJsonArray entrance;
+            QJsonArray multiexit;
+
+            mapJson.insert("width", layer->map()->width());
+            mapJson.insert("height", layer->map()->height());
+
+            mapJson.insert("map", packMapData(layer->asTileLayer()));
+
+            for (int i = 0; i < layer->map()->width(); ++i) {
+                for (int j = 0; j < layer->map()->height(); ++j) {
+                    int tileId = layer->asTileLayer()->cellAt(i, j).tileId();
+
+                    if (tileId < 0) {
+                        mError = tr("Hole in logic layer at (%1, %2)").arg(i).arg(j);
+                        return false;
+                    }
+
+                    switch (tileId) {
+                    case TileId::ENTRANCE:
+                        entrance.append(i);
+                        entrance.append(j);
+                        break;
+                    case TileId::EXIT:
+                    case TileId::LOCKED_EXIT:
+                    case TileId::UNLOCKED_EXIT: {
+                        QJsonArray exit;
+                        exit.append(i);
+                        exit.append(j);
+                        multiexit.append(exit);
+                    } break;
+                    }
+                }
+            }
+
+            mapJson.insert("entrance", entrance);
+            mapJson.insert("multiexit", multiexit);
+        }
+
+        if (layer->name() == "base")
+            mapJson.insert("baseTileVar", packMapData(layer->asTileLayer()));
+
+        if (layer->name() == "deco2")
+            mapJson.insert("deco2TileVar", packMapData(layer->asTileLayer()));
+
+        if (layer->name() == "roof_base")
+            mapJson.insert("roofBaseTileVar", packMapData(layer->asTileLayer()));
+
+        if (layer->name() == "roof_deco")
+            mapJson.insert("roofDecoTileVar", packMapData(layer->asTileLayer()));
+
+        if (layer->name() == "deco") {
+            mapJson.insert("decoTileVar", packMapData(layer->asTileLayer()));
+            mapJson.insert("customTiles", true);
+
+            if (!insertTilesetFile(*layer, "tiles", mapJson))
+                return false;
+
+            QJsonArray decoDesc;
+            QJsonArray decoName;
+
+            auto tilesets = layer->asTileLayer()->usedTilesets();
```

Comment:
Since this function is virtual we don't need to get a `TileLayer*` first.
```suggestion
            auto tilesets = layer->usedTilesets();
```